### PR TITLE
Allow BreakMeshByBlock To save the fake neighbor list to a csv file

### LINKED ISSUE #15970: Allow BreakMeshByBlock to dump fake neighbor cconnectivty

### PR DESCRIPTION
**Pull request by arovinelli**
This PR allows `BreakMeshByBlock` to save the fake neighbor list to file
closes #15970

### CODE REVIEW

**Comment by arovinelli** [comment]
@loganharbour this is ready for review

**Comment by dschwen** [comment]
Why don't you make this restartable data?

**Comment by dschwen** [comment]
Seeing element ids stored that way with a user-exposed mesh generator makes me feel a bit uneasy. We can not rely on element IDs being stable when running with different processor numbers on distributed meshes, can we? And while restart wouldn't work in that scenario anyways, these objects could still be used.

**Comment by arovinelli** [comment]
> Why don't you make this restartable data?

what are restartable data? though using an external list would allow removing some limitations of `BreakMeshByBlock`(doesn't work on distributed meshes). If element id and face order are not guaranteed then I can't figure out anything better.

> Seeing element ids stored that way with a user-exposed mesh generator makes me feel a bit uneasy. We can not rely on element IDs being stable when running with different processor numbers on distributed meshes, can we? And while restart wouldn't work in that scenario anyways, these objects could still be used.

@dschwen The idea behind this PR is to allow users to restart a mesh that has been split using `BreakMeshByBlock` given a list of neighbors in both serial and distributed environments. I can't rely on node position for a restart because the main application I have in mind is cohesive zone modeling. When the mesh gets distorted I can't use geometry to determine initial neighbors (originally coincident nodes will not be coincident after deformation). So my best bet was hoping that once I generate an element side neighbor list given proper ids everything would be fine. If the consistency of nodes id and element ids are not guaranteed then I'm doomed.

Open to better idea, of course

**Comment by arovinelli** [comment]
@dschwen NxN restart should work unless my tests are too small

**Comment by arovinelli** [comment]
Given @dschwen comments, which are totally legitimate but raise questions, we should sort out the stability of elemnt_ids and node numbering for good. Can anyone more confident with MOOSE and libmesh internals chime in. 
I'm thinking to ping @roystgnr, @friedmud and @lindsayad.
If element id and nodes id are not guaranteed to be consistent throughout calculations (including restarts and distributed mesh),  how do we solve an issue in which we need some kind of unique number that cannot be reconstructed using the current geometry?

**Comment by dschwen** [comment]
I'm not saying element IDs are that unstable. I'm saying this design
exposes functionality that should be under the hood and for restart only to
the user, where it can be used in scenarios where element IDs might not be
what the user expects.

**Comment by dschwen** [comment]
You can declare data as restartable by making it a reference in your object and binding it to an object returned by `declareRestartableData`. Framework team members should probably chime in how to best use this in your specific case, as I'm not sure restartable data on mesh generators is even supported. It could very well be that upon restart the mesh is read in and the generators are skipped. 
Interesting that the neighbor pointers are not serialized in the restart...

**Comment by arovinelli** [comment]
@dschwen Generators are not skipped upon restart, there is a [test](https://github.com/idaholab/moose/pull/15972/files#diff-52ba87b6c7037714742cd0e57e04ffcd80c0871c4158ce6efe2c3d492af5c3a0) demonstrating that.
I explicitly save neighbors to a file because `BreakMeshByBlock` doesn't work with checkpoint files #15965.
The procedure for allowing restart with fake neighbors is the following:
- split the mesh `BreakMeshByBlock` with --meh-only and save an exodus of the broken mesh and the neighbor connectivity.
- run your problem using the split exodus and using `AddFakeNeighborsFromList` (which use the csv data). This will allow checkpoint files to work properly.

At any restart, you will still need to add `AddFakeNeighborsFromList` because the fake neighbor's pointers are not saved.

Notice that `AddFakeNeighborsFromList `would also allow to completely skip the `BreakMeshByblock` step if the user has a broken mesh and the list of neighbors. I understand this could go wild easily, especially for first-time users.

In the long run, what I would like to do is to turn`BreakMeshByblock` in a mesh-only tool not available for the actual run and only use `AddFakeNeighborsFromList` in the simulation environment.

Using restartable data might be a good idea, but I have no clue how to save them from a MeshGenerator. Also, this would imply restructuring the restart/relaod procedure to detect and reload fake neighbors. The other options (as you mentioned) would be to save all neighbor pointers. I guess restart experts will not like any of the above options, as `FakeNeighbors` are only a workaround to the lack of cohesive elements in moose and libmesh.

**Comment by roystgnr** [comment]
> If element id and nodes id are not guaranteed to be consistent throughout calculations (including restarts and distributed mesh), how do we solve an issue in which we need some kind of unique number that cannot be reconstructed using the current geometry?

Either turn on the guarantee (```MeshBase::allow_renumbering(false)```), or use some kind of unique number (```DofObject::unique_id()```).

**Comment by roystgnr** [comment]
> I didn't know we have that!

Glad to help!  It's a better solution if you can use it (i.e. if it's on at configure time - which isn't true by default because of the memory cost but which is always true with MOOSE, and if you don't mind that the ids may be too sparse to use as vector indices, so you can only efficiently use them to index into a map or unordered_map).  When we turn off renumbering then our regular ids get more sparse in many cases (e.g. distributed refinement, or *any* kind of adaptive coarsening), which hurts *other* code that is expecting densely numbered ids, for efficiency, or in painful cases like ReplicatedMesh::n_elem(), for correctness.

**Comment by arovinelli** [comment]
> Glad to help! It's a better solution if you can use it (i.e. if it's on at configure time - which isn't true by default because of the memory cost but which is always true with MOOSE, and if you don't mind that the ids may be too sparse to use as vector indices, so you can only efficiently use them to index into a map or unordered_map). When we turn off renumbering then our regular ids get more sparse in many cases (e.g. distributed refinement, or _any_ kind of adaptive coarsening), which hurts _other_ code that is expecting densely numbered ids, for efficiency, or in painful cases like ReplicatedMesh::n_elem(), for correctness.

This case is exactly a map, I don't really mind them being sparse, as long as they are consistent :)

**Comment by arovinelli** [comment]
`DofObject::unique_id()` didn't work, instead I used `ElementInteger` generated on the fly while breaking the mesh.
now the map is correct and should match user expectation (e.g. those numbers will never change)... though the procedure to generate a checkpoint file is still involves 2 steps. Need suggestions here, and still need to explore the restartable data option suggested by @dschwen

**Comment by arovinelli** [comment]
@dschwen also I would like an opinion about changes to `FileMeshGenerator`, is that ok or you would prefer a completely separate mesh generator. Note that to note duplicate code I would still need to modify the current FileMeshGenerator to allow intercepting the mesh creation and temporarily blocking remote element deletion and partitioning. Though I could derive from it

**Comment by dschwen** [comment]
Is `MeshMetaDataInterface` the solution maybe?

**Comment by arovinelli** [comment]
> Is `MeshMetaDataInterface` the solution maybe?

it should be... , but I'm not able to make it work properly (see #16004 )

**Comment by aeslaughter** [comment]
We should make a custom Output object for this instead, such a FakeNeighborList. It can error if the correct information doesn't exist. All tasks related to output should be an Output object. I can help with this if you needed it.

**Comment by arovinelli** [comment]
ok, I tried different ways but I can't make this work. so I'm calling for @aeslaughter help.
This PR + #16004 together should allow me to do what i want.
Let me explain what i need in general.
I need `BreakMeshByBlockGenrator` to store restartable data somewhere, I don't care where as long as such data can be retrieved on restart (e.g. when using a cpr file as input of filemeshgenerator) before preparing the mesh. Data will be a long vector<vector<int>> of integers with dimension N by 4, or std:map<std:pair<elem_id,side>,std:pair<neighbor_id,neighbor_side>>. Note that this data cannot be reconstructed later on by any means.
Once I have this data saved (you need to tell me how to save them), I need a way to check if they are present in the restartable data, and if so I need to retrieve them before preparing the mesh. This is necessary because these data are required to reconstruct some part of the mesh connectivity. If I don't do this before you call prepare_for_use, all ghosting will be wrong.

With the csv the process works like this:
I run my simulation, the meshgenerator does what it needs to do and saves data into a csv file.
On restart, I read that csv from FileMeshGenrator, do my stuff, prepare for use and then run whatever I need to run.
So basically you need to tell me what should I do because RestartableDataIO and MeshMetaDataInterface are really confusing for me.

Thanks in advance for the help

**Comment by aeslaughter** [comment]
It should be possible to do this without writing a CSV file. I will take a look at the PR and see if I can figure it out. I will bug you on slack when I have questions.

**Comment by aeslaughter** [comment]
@arovinelli  Can you fix these conflicts before I pull this branch?

**Comment by arovinelli** [comment]
@aeslaughter done. 
this does not have conflicts but has std::cout here and there, so it will not pass prechecks.
I used the `.sh` file to do the presplit and then run the reload. so you can use that to run the proper chain
